net_connection/src/protocol.rs

```diff
@@ -0,0 +1,276 @@
+//! This module provides the core low-level protocol enumeration
+//! for communications between holochain core and the p2p / networking
+//! process or library. See protocol_wrapper for a higher level interface.
+
+use rmp_serde;
+use serde_bytes;
+
+use holochain_core_types::json::JsonString;
+
+/// Low-level interface spec for communicating with the p2p abstraction
+/// notice this is not Serializable or Deserializable
+/// rmp_serde doesn't serialize enums very well... it uses indexes and arrays
+/// which are not (easily) compatible with other endpoints
+/// we use to/from NamedBinaryData to provide our own serialization wrapper
+#[derive(Debug, Clone, PartialEq)]
+pub enum Protocol {
+    /// send/recv binary data / i.e. encryption, signature messages
+    NamedBinary(NamedBinaryData),
+    /// send/recv generic json as utf8 strings
+    Json(JsonString),
+    /// send/recv a Ping message (ipc protocol spec)
+    Ping(PingData),
+    /// send/recv a Pong message (ipc protocol spec)
+    Pong(PongData),
+    /// we have connected / configured the connection, ready for messages
+    P2pReady,
+}
+
+/// provide utility for Protocol serialization
+impl<'a> From<&'a Protocol> for NamedBinaryData {
+    fn from(p: &'a Protocol) -> Self {
+        match p {
+            Protocol::NamedBinary(nb) => NamedBinaryData {
+                name: b"namedBinary".to_vec(),
+                data: rmp_serde::to_vec_named(nb).unwrap(),
+            },
+            Protocol::Json(j) => NamedBinaryData {
+                name: b"json".to_vec(),
+                data: String::from(j).as_bytes().to_vec(),
```

Comment:
`JsonString` offers `into_bytes()` that gives a `Vec<u8>` of the inner string, should be what you want?

Reply:
:+1: done